Exclude `UPDATE` operations in constraints for immutable fields

ordovicia opened the pull request:
**What this PR does / why we need it**: This PR will fix a problem described in #346, by excluding `UPDATE` operations in constraint templates for immutable fields.

**Which issue(s) does this PR fix** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #346 

<!--
**Are you making changes to Gatekeeper Library policies?**
Please refer to [How to contribute to the library](https://open-policy-agent.github.io/gatekeeper-library/website/#how-to-contribute-to-the-library) to add new policies or update existing policies.

**Are you updating an existing policy?**
Please run `make generate generate-website-docs generate-artifacthub-artifacts` to generate the templates and docs.
-->

**Special notes for your reviewer**: None.

ordovicia commented:
Hi @apeabody @maxsmythe , could you take a look? Thank you!

apeabody reviewed (commented):
Thanks for the contribution @ordovicia!

A few thoughts/questions:
- Just confirming we want/need to exclude `PATCH`?  The default validating webhook verbs are `CREATE` and `UPDATE`: https://github.com/open-policy-agent/gatekeeper/blob/v3.12.0/deploy/gatekeeper.yaml#L3747 | @maxsmythe, any caveats to `PATCH`?
- Thanks for the rego/opa test for `lib.exclude_update_patch`!  We should probably also include some tests for the templates (and maybe even constraints):
  - For the templates you can add `"operation": "UPDATE"` to the `"review": {}`, for example: https://github.com/open-policy-agent/gatekeeper-library/blob/master/src/general/automount-serviceaccount-token/src_test.rego#L10
  - For the constraints you could use a [`AdmissionReview` object](https://open-policy-agent.github.io/gatekeeper/website/docs/gator#usage-1) to include `request.operation: UPDATE` for the test suite.

maxsmythe commented:
PATCH is probably irrelevant. From the K8s docs:

```
Operations is the operations the admission hook cares about - CREATE, UPDATE, DELETE, CONNECT or * for all of those operations and any future admission operations that are added. If '\*' is present, the length of the slice must be one. Required.
```

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#rulewithoperations-v1-admissionregistration-k8s-io

ordovicia commented:
Thank you for the comments!

Fixed to exclude `UPDATE` operation only.
Also added tests for constraint templates.
But I could not find how to add `UPDATE` operation tests for constraints.

apeabody commented:
> Thank you for the comments!
> 
> Fixed to exclude `UPDATE` operation only. Also added tests for constraint templates. But I could not find how to add `UPDATE` operation tests for constraints.

Thanks @ordovicia!  You *should* be able to use a [`AdmissionReview`](https://open-policy-agent.github.io/gatekeeper/website/docs/gator#usage-1) object in the test suite to include the operation type.  For example with `K8sPSPAutomountServiceAccountTokenPod` you could add a new allowed similar to the existing disallowed but allowed for the `UPDATE` operation [test](https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/general/automount-serviceaccount-token/suite.yaml#L15).  Instead of the using the [`Pod`](https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/general/automount-serviceaccount-token/samples/automount-serviceaccount-token/example_disallowed.yaml) object directly, for `object:` you would use an `AdmissionReview` object similar to:

```
kind: AdmissionReview
apiVersion: admission.k8s.io/v1beta1
request:
  operation: "UPDATE"
  object:
    apiVersion: v1
    kind: Pod
    metadata:
      name: nginx-automountserviceaccounttoken-update-allowed
      labels:
        app: nginx-automountserviceaccounttoken
    spec:
      automountServiceAccountToken: true
      containers:
        - name: nginx
          image: nginx
```

The idea being this `Pod` should normally raise a violation, but not if the operation is "UPDATE".

ordovicia commented:
Hi @apeabody @maxsmythe , could you take a look again?

apeabody commented on src/rego/lib_exclude_update/lib_exclude_update.rego:
```diff
@@ -0,0 +1,7 @@
+package lib.exclude_update
+
+import future.keywords.in
```
nit: If we aren't using `in` we should probably drop from all `lib.exclude_update`?

ordovicia replied:
Good catch! I removed all `future.keywords.in` import.

apeabody commented on website/docs/validation/volumes.md:
````diff
@@ -182,6 +196,49 @@ Usage
 ```
 
+</details>
+<details>
````
NOTE ONLY: Opened https://github.com/open-policy-agent/gatekeeper-library/issues/400 as we might want to exclude `AdmissionReview` from being used as documentation samples.

apeabody reviewed (commented):
Thanks @ordovicia!

One question/comment on `lib.exclude_update`.

Cheers!

ordovicia commented:
Hi @apeabody please merge this if it's OK. Thank you!

apeabody commented on library/general/ephemeralstoragelimit/template.yaml:
```diff
@@ -150,10 +151,14 @@ spec:
         }
 
         violation[{"msg": msg}] {
+          not is_update(input.review)
```
Might want to include a comment similar to above.

apeabody commented on library/general/ephemeralstoragelimit/suite.yaml:
```diff
@@ -27,3 +27,7 @@ tests:
+  - name: ephemeral-storage-limit-update
+    object: samples/container-must-have-ephemeral-storage-limit/update.yaml
+    assertions:
+    - violations: no
```
nit:

```suggestion
    - violations: no

```

apeabody commented on library/pod-security-policy/capabilities/suite.yaml:
```diff
@@ -19,3 +19,7 @@ tests:
+      - name: update
+        object: samples/capabilities-demo/update.yaml
+        assertions:
+          - violations: no
```
nit:
```suggestion
          - violations: no

```

apeabody reviewed (approved):
Thanks for the contribution @ordovicia!

Another maintainer will need to review as well.

ritazh reviewed (approved):
LGTM

Thanks for the fix @ordovicia! 🎉  You might want to update the PR description to remove `patch`
